docs/proposals/read-write-operations-bucket.md

```diff
@@ -0,0 +1,248 @@
+
+## Action Items:
+
+*Ensure our block operations are following those rules (e.g create methods that ensures this).
+*Tests edge cases.
+
+## Alternatives:
+
+### Require strong consistent object storage.
+
+As mentioned in #Motivation section this would block community adoption as only a few of object storages has strong or even
+ clearly stated consistency guarantees.
+
+### Full immutability 
+
+By not removing any object, we may simplify problem. However we need hard deletions, because they are:
+- enormously reducing storage size (e.g retention is exactly for this)
+- enormously reducing number of blocks in storage, improving syncing time.
+
+### Special integrity JSON file
+
+We may want to add some file to commit and mark integrity of the block. That file may hold some information like
+files we expect to have in the directory and maybe some hashes of those. 
+
+This was rejected for we can reuse existing meta.json as "integrity" marker (e.g if after 15m we don't have meta file, it is partially uploaded and should be removed).
+
+Additional information like expected number of files can be added later if needed, however with rule no 3. we should not need that.
+
+The main drawback of this is changing or required additional things from TSDB block format which increases complexity.
+
+### Additional coordination based on system with strong consistency (e.g etcd)
+
+As this would help with mentioned issues, Thanos aims for being coordination free and having no additional system dependency to 
+stay operationally simple.
+ 
+### Optimistic/vertical compaction
+
+Currently, compactor requires the blocks to appear in order of creation. That might be not the case for eventual consistent
+ storage. In those cases compactor assumes gap (no block being produced) and compacts with the mentioned gap. Once the new 
+ block is actually available for reading it causes compactor to halt due to overlap. [Vertical compaction](https://github.com/prometheus/tsdb/pull/370) 
+ might help here as can handle those overlaps gracefully, by compacting those together. This however:
+ 
+ * Heavily complicates the compaction
+ * Malformed block in case of simple misconfiguration (misconfigured external labels - quite often happens)
+ * There are lots of edge cases for eventual consistent storage that will cause vertical compaction to be required quite often
+ which will affect performance.
+ * Does not help with partial upload and syncing issues
+ 
+ As vertical compaction might be something we can support in future, it clearly does not help with problems we stated here.
+ 
+### Do not introduce `deleteDelay`
+ 
+We may avoid introducing additional state, by adding mitigation for not having delayed removal.
+
+For example for retention apply or manual block deletion, when we would delete block immediately we can have query failures (object does not exists for getrange operations)
```

Comment:
Is retention apply a know operation? This doesn't quite read right...

Reply:
Yes, and actually the method is called like this. We can reword this.. How would you call it differently?